### PR TITLE
[WIP] Add a Service Book class with a shared testProject method, for scripted pipelines

### PR DESCRIPTION
**Pull request by stephendonner**
@davehunt @tarekziade @rpappalax - thoughts and feedback much-desired and appreciated!

I anticipate a fair amount of cleanup needed/desired here, but this is, IMHO, very much a working prototype.

It can be seen working (i.e. getting to the same "tox" error) on https://qa-preprod-master.fxtest.jenkins.stage.mozaws.net/job/kinto.stage.stephend/15/console as in https://qa-preprod-master.fxtest.jenkins.stage.mozaws.net/job/kinto.stage/3/console

Full output (patch): https://gist.github.com/stephendonner/9a3904b1294c12b9ecc67ea42c5ced71 vs. (existing): https://gist.github.com/stephendonner/146f85214800dd25527974ac52929f5a

Since it's a shared library, we'll also need corresponding changes to call testProject() from a project's Jenkinsfile, which I've done (also rough) in https://github.com/Kinto/kinto-integration-tests/pull/35

### CODE REVIEW

File: src/org/mozilla/fxtest/ServiceBook.groovy [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            sh 'exit 0'
+        } else {
+            sh 'exit 1'
+            }
+        }
+    }
+}
+
+
+return this;
```
Comment (davehunt) [inline src/org/mozilla/fxtest/ServiceBook.groovy]: Is this needed? It looks superfluous.

File: src/org/mozilla/fxtest/ServiceBook.groovy [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    runStage(test)
+                } catch (exc) {
+                    echo test.name + " failed"
+                    echo "Caught: ${exc}"
+                    failures.add(test.name)
+                }
+            }
+        }
+    stage('Ship it!') {
+        if (failures == 0) {
```
Comment (davehunt) [inline src/org/mozilla/fxtest/ServiceBook.groovy]: This should be checking the size of the failures array.

File: src/org/mozilla/fxtest/ServiceBook.groovy [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                sh "${WORKSPACE}/run"
+            }
+        } else {
+            throw new IOException(test.url + " is not allowed")
+        }
+    }
+}
+
+def testProject(String name) {
+    node {
```
Comment (davehunt) [inline src/org/mozilla/fxtest/ServiceBook.groovy]: We might not want to put the `node` here. We're using `node` within the stages, and this is only needed for the exit codes on lins 158-160, so we could simple wrap those in a node. I think that would be a more efficient use of our executors.

File: src/org/mozilla/fxtest/ServiceBook.groovy [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                checkout([$class: 'GitSCM',
+                branches: [[name: '*/master']],
+                doGenerateSubmoduleConfigurations: false,
+                extensions: [[$class: 'CleanCheckout']],
+                submoduleCfg: [],
+                userRemoteConfigs: [[url: test.url + '.git']]]
+                )
+            }
+            echo "checked out"
+            node {
```
Comment (davehunt) [inline src/org/mozilla/fxtest/ServiceBook.groovy]: Note that this will run on any executor. If we have a specific agent or label that we wanted to run on, we should do that. I suspect we want a specific docker image.

File: src/org/mozilla/fxtest/ServiceBook.groovy (hunk 1 above)
Comment (tarekziade) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: If you don't return "this", you can't use the module as a top namespace object in other modules, but maybe there are better ways to do it

Comment (davehunt) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: Hmm.. it was working fine when I had a dummy method in https://github.com/davehunt/fxtest-jenkins-pipeline/blob/servicebook/src/org/mozilla/fxtest/ServiceBook.groovy without `return this;`

Comment (tarekziade) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: maybe that's what "package" is for :)

File: src/org/mozilla/fxtest/ServiceBook.groovy (hunk 4 above)
Comment (stephendonner) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: I don't yet know much about these actual tests, and/or where/how they best should run, besides that they should run in Docker, on the slave/execution node (at least in the qa-preprod* new setup).

Can we defer that for now, or should we hold off on merging this in until we have it clarified and codified, here?

File: src/org/mozilla/fxtest/ServiceBook.groovy [hunk 5]
```diff
@@ -0,0 +1,167 @@
+package org.mozilla.fxtest
+
+import groovy.json.JsonSlurperClassic;
+
+
+HttpResponse doGetHttpRequest(String requestUrl){
```
[the remark this answers is not in the document]
Comment (stephendonner) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: Futured, for now, but will file an Issue for us to get that evaluated.

**Comment by stephendonner** [comment]
Latest run is https://qa-preprod-master.fxtest.jenkins.stage.mozaws.net/job/kinto.stage.stephend/19/console.

I'll work as I have time, now, on version history and code examples, in the README.  Dave and I will likely pair on the tests next week.

File: src/org/mozilla/fxtest/ServiceBook.groovy (hunk 5 above)
[the remark this answers is not in the document]
Comment (stephendonner) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: @davehunt filed https://github.com/mozilla/fxtest-jenkins-pipeline/issues/25, as you've seen; thanks for the reminder/gentle prod :-)

File: src/org/mozilla/fxtest/ServiceBook.groovy (hunk 3 above)
Comment (stephendonner) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: Thanks; fixed!

File: src/org/mozilla/fxtest/ServiceBook.groovy (hunk 2 above)
Comment (stephendonner) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: Fixed, I hope!

File: src/org/mozilla/fxtest/ServiceBook.groovy [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        }
+    }
+    echo "The project " + name + " was not found"
+    return null;
+}
+
+
+
+def validURL(url) {
+    allowedOrgs = ['Kinto', 'mozilla', 'mozilla-services', 'tarekziade', 'rpappalax']
```
Comment (rpappalax) [inline src/org/mozilla/fxtest/ServiceBook.groovy]: This restriction is good to be aware of. I was thinking folks could make sure tests were somewhat viable before pushing them to a mozilla* org (i.e. that tests could be triggered from any org), but this is much safer route.   I suppose we can still follow that workflow in PreProdJenkins and do soak testing there (without hooking into pipeline).
If you don't need the org for tarekziade or me, I'd go ahead and remove them.

Comment (stephendonner) [reply src/org/mozilla/fxtest/ServiceBook.groovy]: Thanks, Richard.  Actually, yes: now that we're getting really close to using this in a production/production-like environment and workflow, it's probably a good time to reel the validation back a bit; I'll make that change now.

File: tests/serviceBookTests.groovy [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    // @Test
+    // void projectWithTests() {
+    //     def resp = [body:'{"data": [{"foo": {"tests": [{"name": "bar", "jenkins_pipeline": true}]}}]}']
+    //     def script = loadScript('src/org/mozilla/fxtest/ServiceBook.groovy')
+    //     script.doGetHttpRequest = { String url -> [resp] }
+    //     def tests = script.getProjectTests('foo')
+    //     assert tests != null
+    // }
```
Comment (davehunt) [inline tests/serviceBookTests.groovy]: Hey @tarekziade, we're trying to mock out the `getProjectTests` but hit various issues with our attempts. Do you have any suggestions for how to test this?

File: README.md [hunk 8]
```diff
@@ -133,6 +133,19 @@ submitToTreeherder(
...
 ```
 
+## testProject
+Queries the Service Book project API for the given project `name`, iterates over
+its associated test repositories, checks them out from SCM, and executes their
+```run``` file(s).  Finally, it returns ```exit 0``` on successful/passing
+tests, and ```exit 1``` in the event of failed builds.
+
+## Examples
+```@Library('fxtest@1.7') _
```
Comment (davehunt) [inline README.md]: This should be <code>```groovy</code> and I don't think we need to include the tag in the example.

File: README.md [hunk 9]
```diff
@@ -133,6 +133,19 @@ submitToTreeherder(
 )
 ```
 
+## testProject
```
Comment (davehunt) [inline README.md]: The other entries are pipeline steps, whereas this is a ServiceBook class. It feels like there should be some separation in the documentation. Perhaps a heading for pipeline steps, and a heading for ServiceBook with the `testProject` method documentated? We're probably close to wanting to break the documentation out to a readthedocs or similar too, as it's getting quite lenghty.

File: README.md [hunk 10]
```diff
@@ -133,6 +133,19 @@ submitToTreeherder(
...
+its associated test repositories, checks them out from SCM, and executes their
+```run``` file(s).  Finally, it returns ```exit 0``` on successful/passing
+tests, and ```exit 1``` in the event of failed builds.
+
+## Examples
+```@Library('fxtest@1.7') _
+
+def sb = new org.mozilla.fxtest.ServiceBook()  
+sb.testProject('kinto')
+
```
Comment (davehunt) [inline README.md]: You need to close the code sample with <code>```</code>, otherwise the next heading is considered code.

File: README.md [hunk 11]
```diff
@@ -158,6 +171,8 @@ writeCapabilities(
 
 ## Version History
 
+### 1.7 (2017-09-04)
+* Introduced `testProject` method, which queries Service Book for project repos.
```
Comment (davehunt) [inline README.md]: I would reword as something like: Introduces `ServiceBook` class, with `testProject` method to execute tests for all pipelines associated with the specified project name.

File: README.md (hunk 10 above)
Comment (stephendonner) [reply README.md]: Fixed; thanks (and, oops!)

File: README.md (hunk 11 above)
Comment (stephendonner) [reply README.md]: Thanks - fixed.

File: README.md (hunk 8 above)
Comment (stephendonner) [reply README.md]: Fixed

File: README.md (hunk 9 above)
Comment (stephendonner) [reply README.md]: Took a stab at this, but currently the "Shared Libraries..." heading and "Pipeline Steps" and "ServiceBook Class" are all the same heading weight of "#" - let me know if that's a deal-breaker, and I'll rework the subheadings' weights to match!

File: README.md [hunk 12]
```diff
@@ -155,9 +156,23 @@ writeCapabilities(
...
+# ServiceBook Class
+## testProject
+Queries the Service Book project API for the given project `name`, iterates over
+its associated test repositories, checks them out from SCM, and executes their
+```run``` file(s).  Finally, it returns ```exit 0``` on successful/passing
+tests, and ```exit 1``` in the event of failed builds.
+
+## Examples
+```groovy
+def sb = new org.mozilla.fxtest.ServiceBook()
```
Comment (davehunt) [inline README.md]: Sorry, I didn't mean to remove the library import, just the version tag as that will become dated.

File: README.md [hunk 13]
```diff
@@ -155,9 +156,23 @@ writeCapabilities(
   path: 'fx51win10.json'
 )
 ```
+# ServiceBook Class
```
Comment (davehunt) [inline README.md]: I'd drop 'Class'

Comment (stephendonner) [reply README.md]: Done

**Comment by stephendonner** [comment]
Filed https://github.com/mozilla/fxtest-jenkins-pipeline/issues/29 to track moving the README to Read the Docs.

File: tests/serviceBookTests.groovy (hunk 7 above)
Comment (tarekziade) [reply tests/serviceBookTests.groovy]: I'll give it a shot
